paasta_tools/async_utils.py

```diff
@@ -10,30 +10,41 @@
 from typing import DefaultDict
 from typing import Dict
 from typing import List
+from typing import Optional
 from typing import TypeVar
 
 
 T = TypeVar("T")
 
 
 def async_ttl_cache(
-    ttl: float = 300, cleanup_self: bool = False
+    ttl: Optional[float] = 300, cleanup_self: bool = False
 ) -> Callable[
     [Callable[..., Awaitable[T]]], Callable[..., Awaitable[T]]  # wrapped  # inner
 ]:
     async def call_or_get_from_cache(cache, async_func, args, kwargs):
         key = functools._make_key(args, kwargs, typed=False)
         try:
             future, last_update = cache[key]
-            if ttl > 0 and time.time() - last_update > ttl:
+            if ttl is not None and time.time() - last_update > ttl:
                 raise KeyError
         except KeyError:
             future = asyncio.ensure_future(async_func(*args, **kwargs))
             # set the timestamp to +infinity so that we always wait on the in-flight request.
             cache[key] = (future, float("Inf"))
-        value = await future
-        cache[key] = (future, time.time())
-        return value
+
+        try:
+            value = await future
+        except Exception:
+            # only update the cache if it's the same future we awaited and
+            # it hasn't already been updated by another coroutine
+            if cache[key] == (future, float("Inf")):
+                del cache[key]
+            raise
+        else:
+            if cache[key] == (future, float("Inf")):
```

Comment:
`cache[key][1]` can be `last_update` or any other valid time from the past, we probably need to compare only the `future` part?

Reply:
Well so when we pull a future out of the cache, if that future has already returned then another coroutine should have already set the timestamp.  If we were to re-set it then we would effectively increase the ttl, which I don't think we want.

If the future we get from the cache has _not_ returned, then the timestamp on the cache entry should still be `float("Inf")`, in which case the first coroutine to return from awaiting the future will update the timestamp to `time.time()`, which starts the ttl count down.''

For the case where the future raises an Exception, I think you are right that we could get away with only looking at the future, but I figured this keeps the logic more consistent.